[Feat] 법정 점검 게시글에 대한 조회 기능을 구현한다

Galmaeki opened the pull request:
# Pull Request 요약

- 월별 요약을 제외한 조회 기능에 대한 API입니다

## 변경 사항

- 이 PR의 코드 주요 내용은 무엇인가요?

## 관련 이슈

- 이 PR은 어떤 이슈로부터 만들어졌나요? (예: `#123`)

## 개발 유형

- [ ] 버그 수정
- [x] 새로운 기능 개발
- [ ] 리팩토링
- [ ] 테스트 코드 작성
- [ ] 문서 업데이트
- 기타 (아래에 설명 추가)

## 작업 기간

- 작업 시작일: (YYYY-MM-DD)
- 작업 종료일: (YYYY-MM-DD)

## 추가 사항
- #143

Galmaeki committed:
[refactor] ScheduleConverter 위치 변경

변경 사항 :
- ScheduleConverter : 패키지 위치 변경

Galmaeki committed:
[feat] 조회에 필요한 Dto 클래스 작성

변경 사항 :
- 조회 시 필요한 ResponseDto 레코드 클래스 작성

Galmaeki committed:
[refactor] 이슈 생성시, 게시글에도 조회하도록 수정

변경 사항 :
- ArticleIssueService : 생성로직 추가
- ManageArticle : 새 필드 추가

Galmaeki committed:
[feat] 필요한 메소드 생성

변경 사항 :
- ManageArticleRepository :
  - officeId 조회 메소드 추가
  - id 조회 메소드 추가

Galmaeki committed:
[feat] 조회 로직 작성

변경 사항 :
- ManageArticleReadController :
  - 요약, 상세, 목록 조회 추가
- ReadManageArticleService :
  - 요약, 상세, 목록 조회 추가
  - summaryArticleFrom, detailArticleFrom, listArticleFrom 메소드 추가

asd42270 commented on src/main/java/com/final_10aeat/domain/manageArticle/service/ReadManageArticleService.java:
```diff
@@ -0,0 +1,115 @@
+            .orElseThrow(ArticleNotFoundException::new);
+
+        if (!article.getOffice().getId().equals(userOfficeId)) {
+            throw new UnauthorizedAccessException();
```
Office에 대한 잘못된 접근이라는 예외를 던지는 건가요??? 조금 더 명시적으로 바꾸는 것은 어떨지 싶습니다...ㅎㅎ...

Galmaeki replied:
조회하려는 게시글의 id가 소유한 office와 맞지 않을 때 발생하는 예외로는 접근 권한이 없다는 에러를 보내는게 가장 적절하다고 생각했습니다!
여기서 더 구체적으로 던질 메시지나 예외가 생각이 나질 않슴다 😢
추천해주실 수 있을까요?

ghrltjdtprbs replied:
저도 UnauthorizedAccessException 던지긴 했습니다! 인가 관련 예외라 이 정도면 충분할 것 같아서여

asd42270 replied:
어머 이제 봤슴다 아하 그러면 괜찮을 것 같아요 ㅋㅋㅋ....하핫

asd42270 reviewed (approved):
고생하셨습니다!! 질문 하나 있어서 남겼습니다!

Galmaeki committed:
[refactor] 클래스 명 변경

변경 사항 :
- ManageArticleReadController -> ReadManageArticleController 로 이름 변경

Galmaeki committed:
[refactor] 검색 성능을 위한 year, month 추가

변경 사항 :
- ManageSchedule : 컬럼 추가

Galmaeki committed:
[refactor] 검색 성능을 위한 year, month 추가

변경 사항 :
- ScheduleConverter : 생성로직 변경

Galmaeki committed:
[feat] 월별, 연도별 검색 기능 추가

변경 사항 :
- ReadManageArticleController :
  - monthlySummary, monthlyListArticle 메소드 추가
- ReadManageArticleService :
  - monthlySummary,listArticleMonthlyNullFrom, listArticleMonthlyFrom 메소드 추가

ghrltjdtprbs commented on src/main/java/com/final_10aeat/domain/manageArticle/dto/response/ManageScheduleResponse.java:
```diff
@@ -0,0 +1,14 @@
+@Builder
+public record ManageScheduleResponse(
+    Long manageScheduleId,
+    Boolean isComplete,
```
boolean 정도는 기본형으로 써도 될 것 같습니다! 😊

Galmaeki replied:
여기선 확실히 기본형을 사용하는게 메모리에서 이득을 볼 수 있을것같슴다 👍

ghrltjdtprbs commented on src/main/java/com/final_10aeat/domain/manageArticle/service/ReadManageArticleService.java:
```diff
@@ -0,0 +1,175 @@
+                articleId)
+            .orElseThrow(ArticleNotFoundException::new);
+
+        if (!article.getOffice().getId().equals(userOfficeId)) {
```
이 메서드에만 officeId 검증 로직을 추가한 이유가 있을까요?

Galmaeki replied:
다른 부분은 articleId를 직접적으로 호출하는 부분이 없기에 회원 정보에 있는 officeId 만으로 정보를 내려주면 되니 검증할 필요가 없지만 이 메소드는 articleId를 직접적으로 호출하기 때문에 호출한 articleId가 회원정보의 office와 다를 가능성이 존재합니다!

ghrltjdtprbs reviewed (approved):
수고하셨습니다! 법정점검 조회 관리자용 api는 따로 만드시는건가요?? 제 생각으로는 법정점검 부분은 관리자나 소유주나 비슷한 정보를 response하기 때문에 합치는 것도 좋은 방법일 것 같습니다!

Galmaeki committed:
[refactor] 필수로 Wrapper 클래스가 아니어도 되는 항목 변경

변경 사항 :
- ManageScheduleResponse : isComplete Wrapper -> Primitive

ghrltjdtprbs commented on src/main/java/com/final_10aeat/domain/manageArticle/service/ReadManageArticleService.java:
monthlyListArticle, listArticleMonthlyFrom, listArticleMonthlyNullFrom 메소드들은 유사한 필터링 로직을 공유하는 것 같네요! 하나의 메소드로 통합하여 코드 중복을 줄이는건 어떨까요??
예를 들어, month 매개변수가 null일 때와 그렇지 않을 때의 로직을 한 메서드 내에서 처리가 가능합니다! 이렇게 하면 ofNullable을 사용하여 필터링 조건을 동적으로 설정할 수 있고, 코드의 중복을 크게 줄일 수 있을 것 같아요!! 😊😊
```java
public List<ListManageArticleResponse> listArticleByMonth(
    Long userOfficeId, Integer year, Optional<Integer> monthOpt
) {
    List<ManageArticle> articles = manageArticleRepository
        .findAllByOfficeIdAndDeletedAtNull(userOfficeId);

    return articles.stream()
        .filter(article -> article.getSchedules().stream()
            .anyMatch(schedule ->
                schedule.getYear().equals(year) &&
                monthOpt.map(month -> schedule.getMonth().equals(month)).orElse(true)
            )
        )
        .map(this::listArticleFrom)
        .toList();
}
```

Galmaeki commented:
> Reference in new issue
> 
> ### Reference in new issue
> Repository
> 10aeat_Backend
> Title

관리자용 api는 readController 내부에 메소드에만 PreAuthorized을 붙여 사용할 예정임다
조회부분에서 D-XX 같은 부분이 필요해서 응답은 다르게 나가야 할것 같슴다 😢

Galmaeki committed:
[style] 주석 변경

변경 사항 :
- 필요하지 않은 주석 변경

ghrltjdtprbs commented on src/main/java/com/final_10aeat/domain/manageArticle/service/ReadManageArticleService.java:
스트림 연산에서 Collectors를 사용하는건 어떨까요?? 복잡한 데이터 변환을 단순화할 수 있습니다 ㅎㅎ  또한, 기존의 for-loop 대신 스트림을 사용하면 병렬 처리도 쉽게 구현할 수 있고 코드도 더 선언적으로 작성할 수 있습니다!
```java
List<Long> issueIds = articles.stream()
    .map(ManageArticle::getIssue)
    .filter(Objects::nonNull)
    .map(ArticleIssue::getId)
    .collect(Collectors.toList());
```

ghrltjdtprbs commented on src/main/java/com/final_10aeat/domain/manageArticle/service/ReadManageArticleService.java:
크게 중요한 부분은 아니지만, 한 메서드가 여러 역할을 가지고 있는 부분이 좀 보이네요. 큰 메소드는 여러 작은 메소드로 나누어 작성하는 것이 좋습니다!(단일 책임의 원칙)👍
예를 들어 summaryArticleFrom 메소드에서는 각 통계 계산과 이슈 ID 추출을 별도의 메소드로 분리할 수 있습니다.
